### PR TITLE
read from cache to deal with streamed bodies correctly

### PR DESCRIPTION
**Pull request by grosser**
streamed bodies can only be read once, so we cannot read them and then return them
hotfix for 1.6.0 release ... need a bit more time and a minor release to implement a proper fix

reverts https://github.com/rtomayko/rack-cache/pull/133
fixes https://github.com/rtomayko/rack-cache/issues/134

review plz: @alau @amatriain @krisdigital 

TODO:
make write return the read body and then return that, except for cases where that might be inefficient like Disk store

### CODE REVIEW

File: lib/rack/cache/meta_store.rb [hunk 1]
```diff
@@ -64,6 +64,18 @@ def store(request, response, entity_store)
...
+        unless entity_store.is_a? Rack::Cache::EntityStore::Noop
+          # A stream body can only be read once and is currently closed by #write.
+          # (To avoid having to keep giant objects in memory when writing to disk cache
+          # the body is never converted to a single string)
+          # We cannot always reply on body to be re-readable,
+          # so we have to read it from the cache.
+          # BUG: if the cache was unable to store a stream, the stream will be closed
+          #      and rack will try to read it again, resulting in hard to track down exception
+          response.body = entity_store.open(digest) || response.body
+        end
```
Comment (amatriain) [inline lib/rack/cache/meta_store.rb]: What about replacing the whole conditional with this?

```
stored_body = entity_store.open(digest)
response.body = stored_body unless stored_body.nil? || stored_body.size.zero? 
```

This would remove checking for the Noop entitystore, which is ugly and imho should not be here.

```unless stored_body.nil? || stored_body.size.zero?```  is the simplest one-liner I can think of that works both with Array and File objects (```Rack::Cache::EntityStore::Disk::Body``` extends ```File```). It would preserve the original response body if it cannot be read from the cache, and Noop would keep working.

File: test/meta_store_test.rb [hunk 2]
```diff
@@ -15,7 +15,6 @@ def mock_request(uri, opts)
 
       def mock_response(status, headers, body)
         headers ||= {}
-        body = Array(body).compact
```
Comment (amatriain) [inline test/meta_store_test.rb]: Why did you remove this? Cleanup, or is it related to the problem with streamable bodies?

Comment (grosser) [inline test/meta_store_test.rb]: cleanup and since I had to pass in an IO object

On Thu, Feb 11, 2016 at 3:13 AM, Alfredo Amatriain <notifications@github.com

File: lib/rack/cache/meta_store.rb (hunk 1 above)
Comment (grosser) [inline lib/rack/cache/meta_store.rb]: we have to support empty bodies too, for now I think the Noop check is the safest, I want to clean this up to make write return the written body so this re-reading goes away ...

**Comment by grosser** [comment]
1.6.1
